src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs

```diff
@@ -108,11 +108,50 @@ public static class TagHelperBlockRewriter
             return attributes;
         }
 
-        private static bool IsSelfClosing(Block beginTagBlock)
+        private static TagMode GetTagMode(
+            string tagName,
+            Block beginTagBlock,
+            IEnumerable<TagHelperDescriptor> descriptors,
+            ErrorSink errorSink)
         {
+            TagHelperDescriptor baseDescriptor = null;
+            foreach (var descriptor in descriptors)
+            {
+                if (descriptor.TagStructure != TagStructure.Unspecified)
+                {
+                    // Can't have a set of TagHelpers that expect different structures.
```

Comment:
Other errors in this file relate to the syntax of the Razor file. This class should continue to assume the descriptors are consistent.

Suggest this code should use the `FirstOrDefault()` descriptor with `TagStructure != TagStructure.Unspecified`. The error about conflicting descriptors for the current element should be in `TgHelperParseTreeRewriter`, just after determining `GetDescriptors()` returned a non-empty set. Might be a bit more complex but that location makes more sense and we could eventually optimize things e.g. tracking descriptor sets and their associated errors.